Python 3.13.0b1: testBasicCase2 and testMutuallyExclusiveArgs fail because the output strings changed values

The help output in Python 3.13 has changed. Now, the compared strings don't match exactly, hence test failures.


```
_______________________ TestBasicUseCases.testBasicCase2 _______________________

self = <tests.test_configargparse.TestBasicUseCases testMethod=testBasicCase2>
use_groups = False

    def testBasicCase2(self, use_groups=False):
    
        ## Test command line, config file and env var values
        default_config_file = tempfile.NamedTemporaryFile(mode="w", delete=False)
        default_config_file.flush()
    
        p = self.initParser(default_config_files=['/etc/settings.ini',
                '/home/jeff/.user_settings', default_config_file.name])
        p.add_arg('vcf', nargs='+', help='Variant file(s)')
        if not use_groups:
            self.add_arg('--genome', help='Path to genome file', required=True)
            self.add_arg('-v', dest='verbose', action='store_true')
            self.add_arg('-g', '--my-cfg-file', required=True,
                         is_config_file=True)
            self.add_arg('-d', '--dbsnp', env_var='DBSNP_PATH')
            self.add_arg('-f', '--format',
                         choices=["BED", "MAF", "VCF", "WIG", "R"],
                         dest="fmt", metavar="FRMT", env_var="OUTPUT_FORMAT",
                         default="BED")
        else:
            g = p.add_argument_group(title="g1")
            g.add_arg('--genome', help='Path to genome file', required=True)
            g.add_arg('-v', dest='verbose', action='store_true')
            g.add_arg('-g', '--my-cfg-file', required=True,
                      is_config_file=True)
            g = p.add_argument_group(title="g2")
            g.add_arg('-d', '--dbsnp', env_var='DBSNP_PATH')
            g.add_arg('-f', '--format',
                      choices=["BED", "MAF", "VCF", "WIG", "R"],
                      dest="fmt", metavar="FRMT", env_var="OUTPUT_FORMAT",
                      default="BED")
    
        # make sure required args are enforced
        self.assertParseArgsRaises("too few arg"
                                   if sys.version_info.major < 3 else
                                   "the following arguments are required: vcf, -g/--my-cfg-file",
                                   args="--genome hg19")
        self.assertParseArgsRaises("Unable to open config file: file.txt. Error: No such file or director", args="-g file.txt")
    
        # check values after setting args on command line
        config_file2 = tempfile.NamedTemporaryFile(mode="w", delete=False)
        config_file2.flush()
    
        ns = self.parse(args="--genome hg19 -g %s bla.vcf " % config_file2.name)
        self.assertEqual(ns.genome, "hg19")
        self.assertEqual(ns.verbose, False)
        self.assertIsNone(ns.dbsnp)
        self.assertEqual(ns.fmt, "BED")
        self.assertListEqual(ns.vcf, ["bla.vcf"])
    
        self.assertRegex(self.format_values(),
            'Command Line Args:   --genome hg19 -g [^\\s]+ bla.vcf\n'
            'Defaults:\n'
            '  --format: \\s+ BED\n')
    
        # check precedence: args > env > config > default using the --format arg
        default_config_file.write("--format MAF")
        default_config_file.flush()
        ns = self.parse(args="--genome hg19 -g %s f.vcf " % config_file2.name)
        self.assertEqual(ns.fmt, "MAF")
        self.assertRegex(self.format_values(),
            'Command Line Args:   --genome hg19 -g [^\\s]+ f.vcf\n'
            'Config File \\([^\\s]+\\):\n'
            '  --format: \\s+ MAF\n')
    
        config_file2.write("--format VCF")
        config_file2.flush()
        ns = self.parse(args="--genome hg19 -g %s f.vcf " % config_file2.name)
        self.assertEqual(ns.fmt, "VCF")
        self.assertRegex(self.format_values(),
            'Command Line Args:   --genome hg19 -g [^\\s]+ f.vcf\n'
            'Config File \\([^\\s]+\\):\n'
            '  --format: \\s+ VCF\n')
    
        ns = self.parse(env_vars={"OUTPUT_FORMAT":"R", "DBSNP_PATH":"/a/b.vcf"},
            args="--genome hg19 -g %s f.vcf " % config_file2.name)
        self.assertEqual(ns.fmt, "R")
        self.assertEqual(ns.dbsnp, "/a/b.vcf")
        self.assertRegex(self.format_values(),
            'Command Line Args:   --genome hg19 -g [^\\s]+ f.vcf\n'
            'Environment Variables:\n'
            '  DBSNP_PATH: \\s+ /a/b.vcf\n'
            '  OUTPUT_FORMAT: \\s+ R\n')
    
        ns = self.parse(env_vars={"OUTPUT_FORMAT":"R", "DBSNP_PATH":"/a/b.vcf",
                                  "ANOTHER_VAR":"something"},
            args="--genome hg19 -g %s --format WIG f.vcf" % config_file2.name)
        self.assertEqual(ns.fmt, "WIG")
        self.assertEqual(ns.dbsnp, "/a/b.vcf")
        self.assertRegex(self.format_values(),
            'Command Line Args:   --genome hg19 -g [^\\s]+ --format WIG f.vcf\n'
            'Environment Variables:\n'
            '  DBSNP_PATH: \\s+ /a/b.vcf\n')
    
        if not use_groups:
>           self.assertRegex(self.format_help(),
                'usage: .* \\[-h\\] --genome GENOME \\[-v\\] -g MY_CFG_FILE\n?'
                '\\s+\\[-d DBSNP\\]\\s+\\[-f FRMT\\]\\s+vcf \\[vcf ...\\]\n\n'
                'positional arguments:\n'
                '  vcf \\s+ Variant file\\(s\\)\n\n'
                '%s:\n'
                '  -h, --help \\s+ show this help message and exit\n'
                '  --genome GENOME \\s+ Path to genome file\n'
                '  -v\n'
                '  -g MY_CFG_FILE, --my-cfg-file MY_CFG_FILE\n'
                '  -d DBSNP, --dbsnp DBSNP\\s+\\[env var: DBSNP_PATH\\]\n'
                '  -f FRMT, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n\n'%OPTIONAL_ARGS_STRING +
                7*r'(.+\s*)')
E           AssertionError: Regex didn't match: 'usage: .* \\[-h\\] --genome GENOME \\[-v\\] -g MY_CFG_FILE\n?\\s+\\[-d DBSNP\\]\\s+\\[-f FRMT\\]\\s+vcf \\[vcf ...\\]\n\npositional arguments:\n  vcf \\s+ Variant file\\(s\\)\n\noptions:\n  -h, --help \\s+ show this help message and exit\n  --genome GENOME \\s+ Path to genome file\n  -v\n  -g MY_CFG_FILE, --my-cfg-file MY_CFG_FILE\n  -d DBSNP, --dbsnp DBSNP\\s+\\[env var: DBSNP_PATH\\]\n  -f FRMT, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n\n(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)' not found in "usage: pytest-3.13 [-h] --genome GENOME [-v] -g MY_CFG_FILE [-d DBSNP]\n                   [-f FRMT]\n                   vcf [vcf ...]\n\npositional arguments:\n  vcf                   Variant file(s)\n\noptions:\n  -h, --help            show this help message and exit\n  --genome GENOME       Path to genome file\n  -v\n  -g, --my-cfg-file MY_CFG_FILE\n  -d, --dbsnp DBSNP     [env var: DBSNP_PATH]\n  -f, --format FRMT     [env var: OUTPUT_FORMAT]\n\nArgs that start with '--' can also be set in a config file (/etc/settings.ini\nor /home/jeff/.user_settings or /tmp/tmpew58xqit or specified via -g). Config\nfile syntax allows: key=value, flag=true, stuff=[a,b,c] (for details, see\nsyntax at https://goo.gl/R74nmi). In general, command-line values override\nenvironment variables which override config file values which override\ndefaults.\n"

tests/test_configargparse.py:265: AssertionError
_________________ TestBasicUseCases.testBasicCase2_WithGroups __________________

self = <tests.test_configargparse.TestBasicUseCases testMethod=testBasicCase2_WithGroups>

    def testBasicCase2_WithGroups(self):
>       self.testBasicCase2(use_groups=True)

tests/test_configargparse.py:305: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_configargparse.py:279: in testBasicCase2
    self.assertRegex(self.format_help(),
E   AssertionError: Regex didn't match: 'usage: .* \\[-h\\] --genome GENOME \\[-v\\] -g MY_CFG_FILE\n?\\s+\\[-d DBSNP\\]\\s+\\[-f FRMT\\]\\s+vcf \\[vcf ...\\]\n\npositional arguments:\n  vcf \\s+ Variant file\\(s\\)\n\noptions:\n  -h, --help \\s+ show this help message and exit\n\ng1:\n  --genome GENOME \\s+ Path to genome file\n  -v\n  -g MY_CFG_FILE, --my-cfg-file MY_CFG_FILE\n\ng2:\n  -d DBSNP, --dbsnp DBSNP\\s+\\[env var: DBSNP_PATH\\]\n  -f FRMT, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n\n(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)' not found in "usage: pytest-3.13 [-h] --genome GENOME [-v] -g MY_CFG_FILE [-d DBSNP]\n                   [-f FRMT]\n                   vcf [vcf ...]\n\npositional arguments:\n  vcf                   Variant file(s)\n\noptions:\n  -h, --help            show this help message and exit\n\ng1:\n  --genome GENOME       Path to genome file\n  -v\n  -g, --my-cfg-file MY_CFG_FILE\n\ng2:\n  -d, --dbsnp DBSNP     [env var: DBSNP_PATH]\n  -f, --format FRMT     [env var: OUTPUT_FORMAT]\n\nArgs that start with '--' can also be set in a config file (/etc/settings.ini\nor /home/jeff/.user_settings or /tmp/tmpaaw7d7xa or specified via -g). Config\nfile syntax allows: key=value, flag=true, stuff=[a,b,c] (for details, see\nsyntax at https://goo.gl/R74nmi). In general, command-line values override\nenvironment variables which override config file values which override\ndefaults.\n"
_________________ TestBasicUseCases.testMutuallyExclusiveArgs __________________

self = <tests.test_configargparse.TestBasicUseCases testMethod=testMutuallyExclusiveArgs>

    def testMutuallyExclusiveArgs(self):
        config_file = tempfile.NamedTemporaryFile(mode="w", delete=False)
    
        p = self.parser
        g = p.add_argument_group(title="group1")
        g.add_arg('--genome', help='Path to genome file', required=True)
        g.add_arg('-v', dest='verbose', action='store_true')
    
        g = p.add_mutually_exclusive_group(required=True)
        g.add_arg('-f1', '--type1-cfg-file', is_config_file=True)
        g.add_arg('-f2', '--type2-cfg-file', is_config_file=True)
    
        g = p.add_mutually_exclusive_group(required=True)
        g.add_arg('-f', '--format', choices=["BED", "MAF", "VCF", "WIG", "R"],
                     dest="fmt", metavar="FRMT", env_var="OUTPUT_FORMAT",
                     default="BED")
        g.add_arg('-b', '--bam', dest='fmt', action="store_const", const="BAM",
                  env_var='BAM_FORMAT')
    
        ns = self.parse(args="--genome hg19 -f1 %s --bam" % config_file.name)
        self.assertEqual(ns.genome, "hg19")
        self.assertEqual(ns.verbose, False)
        self.assertEqual(ns.fmt, "BAM")
    
        ns = self.parse(env_vars={"BAM_FORMAT" : "true"},
                        args="--genome hg19 -f1 %s" % config_file.name)
        self.assertEqual(ns.genome, "hg19")
        self.assertEqual(ns.verbose, False)
        self.assertEqual(ns.fmt, "BAM")
        self.assertRegex(self.format_values(),
            'Command Line Args:   --genome hg19 -f1 [^\\s]+\n'
            'Environment Variables:\n'
            '  BAM_FORMAT: \\s+ true\n'
            'Defaults:\n'
            '  --format: \\s+ BED\n')
    
>       self.assertRegex(self.format_help(),
            r'usage: .* \[-h\] --genome GENOME \[-v\]\s+ \(-f1 TYPE1_CFG_FILE \|'
            ' \\s*-f2 TYPE2_CFG_FILE\\)\\s+\\(-f FRMT \\| -b\\)\n\n'
            '%s:\n'
            '  -h, --help            show this help message and exit\n'
            '  -f1 TYPE1_CFG_FILE, --type1-cfg-file TYPE1_CFG_FILE\n'
            '  -f2 TYPE2_CFG_FILE, --type2-cfg-file TYPE2_CFG_FILE\n'
            '  -f FRMT, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n'
            '  -b, --bam\\s+\\[env var: BAM_FORMAT\\]\n\n'
            'group1:\n'
            '  --genome GENOME       Path to genome file\n'
            '  -v\n\n'%OPTIONAL_ARGS_STRING +
            5*r'(.+\s*)')
E       AssertionError: Regex didn't match: 'usage: .* \\[-h\\] --genome GENOME \\[-v\\]\\s+ \\(-f1 TYPE1_CFG_FILE \\| \\s*-f2 TYPE2_CFG_FILE\\)\\s+\\(-f FRMT \\| -b\\)\n\noptions:\n  -h, --help            show this help message and exit\n  -f1 TYPE1_CFG_FILE, --type1-cfg-file TYPE1_CFG_FILE\n  -f2 TYPE2_CFG_FILE, --type2-cfg-file TYPE2_CFG_FILE\n  -f FRMT, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n  -b, --bam\\s+\\[env var: BAM_FORMAT\\]\n\ngroup1:\n  --genome GENOME       Path to genome file\n  -v\n\n(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)(.+\\s*)' not found in "usage: pytest-3.13 [-h] --genome GENOME [-v]\n                   (-f1 TYPE1_CFG_FILE | -f2 TYPE2_CFG_FILE) (-f FRMT | -b)\n\noptions:\n  -h, --help            show this help message and exit\n  -f1, --type1-cfg-file TYPE1_CFG_FILE\n  -f2, --type2-cfg-file TYPE2_CFG_FILE\n  -f, --format FRMT     [env var: OUTPUT_FORMAT]\n  -b, --bam             [env var: BAM_FORMAT]\n\ngroup1:\n  --genome GENOME       Path to genome file\n  -v\n\nArgs that start with '--' can also be set in a config file (specified via -f1\nor -f2). Config file syntax allows: key=value, flag=true, stuff=[a,b,c] (for\ndetails, see syntax at https://goo.gl/R74nmi). In general, command-line values\noverride environment variables which override config file values which\noverride defaults.\n"

tests/test_configargparse.py:385: AssertionError
FAILED tests/test_configargparse.py::TestBasicUseCases::testBasicCase2 - Asse...
FAILED tests/test_configargparse.py::TestBasicUseCases::testBasicCase2_WithGroups
FAILED tests/test_configargparse.py::TestBasicUseCases::testMutuallyExclusiveArgs
============ 3 failed, 32 passed, 15 deselected, 1 warning in 0.18s ============
```

--- a/.github/workflows/apidocs.yml
+++ b/.github/workflows/apidocs.yml
@@ -13,9 +13,7 @@ jobs:
   deploy:
     runs-on: macos-latest
     permissions:
-      contents: read
-      pages: write
-      id-token: write
+      contents: write
 
     steps:
       - uses: actions/checkout@v4
@@ -33,7 +31,7 @@ jobs:
         run: ./apidocs.sh
 
       - name: Push API documentation to Github Pages
-        uses: peaceiris/actions-gh-pages@373f7f263a76c20808c831209c920827a82a2847 # v3.9.3
+        uses: peaceiris/actions-gh-pages@4f9cc6602d3f66b9c108549d475ec49e8ef4d45e   # v4.0.0
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
           publish_dir: ./apidocs

--- a/configargparse.py
+++ b/configargparse.py
@@ -19,7 +19,6 @@ from collections import OrderedDict
 import textwrap
 from io import StringIO
 
-
 ACTION_TYPES_THAT_DONT_NEED_A_VALUE = [
     argparse._StoreTrueAction,
     argparse._StoreFalseAction,
@@ -516,13 +515,27 @@ class TomlConfigParser(ConfigFileParser):
 
     def parse(self, stream):
         """Parses the keys and values from a TOML config file."""
-        # parse with configparser to allow multi-line values
-        import toml
-
+        # Use tomllib (Python 3.11+) if available, otherwise fall back to toml package
         try:
-            config = toml.load(stream)
-        except Exception as e:
-            raise ConfigFileParserException("Couldn't parse TOML file: %s" % e)
+            import tomllib
+
+            # tomllib.load() requires binary mode, so use loads() for stream compatibility
+            try:
+                content = stream.read()
+                # If content is bytes, decode it; if string, use as-is
+                if isinstance(content, bytes):
+                    content = content.decode("utf-8")
+                config = tomllib.loads(content)
+            except Exception as e:
+                raise ConfigFileParserException("Couldn't parse TOML file: %s" % e)
+        except ImportError:
+            # Fall back to toml package (supports text mode)
+            import toml
+
+            try:
+                config = toml.load(stream)
+            except Exception as e:
+                raise ConfigFileParserException("Couldn't parse TOML file: %s" % e)
 
         # convert to dict and filter based on section names
         result = OrderedDict()
@@ -549,6 +562,40 @@ class TomlConfigParser(ConfigFileParser):
             "Config file syntax is Tom's Obvious, Minimal Language. "
             "See https://github.com/toml-lang/toml/blob/v0.5.0/README.md for details."
         )
+
+    def serialize(self, items):
+        """Serialize items to TOML format with section support.
+
+        Note: Requires the 'toml' package for serialization (pip install toml).
+        Python 3.11's tomllib only supports reading, not writing.
+        """
+        # lazy-import to avoid dependency unless class is used
+        try:
+            import toml
+        except ImportError:
+            raise ConfigFileParserException(
+                "The 'toml' package is required for TOML serialization. "
+                "Install it with: pip install toml"
+            )
+
+        # Put items in the first configured section
+        if not self.sections:
+            # No sections configured, serialize as flat TOML
+            return toml.dumps(dict(items))
+
+        # Create nested dict structure for section
+        section_name = self.sections[0]
+        sections = parse_toml_section_name(section_name)
+
+        # Build nested dict from section path
+        result = {}
+        current = result
+        for i, section in enumerate(sections[:-1]):
+            current[section] = {}
+            current = current[section]
+        current[sections[-1]] = dict(items)
+
+        return toml.dumps(result)
 
 
 class IniConfigParser(ConfigFileParser):
@@ -598,7 +645,7 @@ class IniConfigParser(ConfigFileParser):
     def __init__(self, sections, split_ml_text_to_list):
         """
         :param sections: The section names bounded to the new parser.
-        :param split_ml_text_to_list: Wether to convert multiline strings to list
+        :param split_ml_text_to_list: Whether to convert multiline strings to list
         """
         super().__init__()
         self.sections = sections
@@ -671,6 +718,37 @@ class IniConfigParser(ConfigFileParser):
             )
         return msg
 
+    def serialize(self, items):
+        """Serialize items to INI format with section support."""
+        config = configparser.ConfigParser()
+
+        # Use the first configured section
+        section_name = self.sections[0] if self.sections else "DEFAULT"
+
+        # Add section if not DEFAULT
+        if section_name != "DEFAULT" and section_name != configparser.DEFAULTSECT:
+            config.add_section(section_name)
+
+        # Add items to the section
+        for key, value in items.items():
+            if isinstance(value, list):
+                # Handle lists
+                if self.split_ml_text_to_list:
+                    # Multi-line format
+                    config.set(
+                        section_name, key, "\n" + "\n".join(str(v) for v in value)
+                    )
+                else:
+                    # Python list syntax
+                    config.set(section_name, key, str(value))
+            else:
+                config.set(section_name, key, str(value))
+
+        stream = StringIO()
+        config.write(stream)
+        stream.seek(0)
+        return stream.read()
+
 
 class CompositeConfigParser(ConfigFileParser):
     """
@@ -689,12 +767,22 @@ class CompositeConfigParser(ConfigFileParser):
 
     def parse(self, stream):
         errors = []
-        for p in self.parsers:
+        for i, p in enumerate(self.parsers):
             try:
                 return p.parse(stream)  # type: ignore[no-any-return]
             except Exception as e:
-                stream.seek(0)
                 errors.append(e)
+                # Try to seek back to beginning for next parser
+                # If this is not the last parser and seek fails, we can't continue
+                if i < len(self.parsers) - 1:
+                    try:
+                        stream.seek(0)
+                    except (AttributeError, OSError):
+                        # Stream doesn't support seeking
+                        raise ConfigFileParserException(
+                            f"Error parsing config with {p.__class__.__name__}: {e}. "
+                            f"Cannot try additional parsers because stream is not seekable."
+                        ) from e
         raise ConfigFileParserException(
             f"Error parsing config: {', '.join(repr(str(e)) for e in errors)}"
         )
@@ -714,6 +802,15 @@ class CompositeConfigParser(ConfigFileParser):
         for i, parser in enumerate(self.parsers):
             msg += f"[{i+1}] {guess_format_name(parser.__class__.__name__)}: {parser.get_syntax_description()} \n"
         return msg
+
+    def serialize(self, items):
+        """Serialize items using the first parser in the composite."""
+        if not self.parsers:
+            raise ConfigFileParserException(
+                "No parsers configured in CompositeConfigParser"
+            )
+        # Use the first parser to serialize
+        return self.parsers[0].serialize(items)  # type: ignore[no-any-return]
 
 
 # used while parsing args to keep track of where they came from
@@ -844,7 +941,8 @@ class ArgumentParser(argparse.ArgumentParser):
                 is_write_out_config_file_arg=True,
             )
 
-        # TODO: delete me!
+        # Workaround for Python < 3.9: exit_on_error parameter was added in 3.9
+        # This can be removed when minimum Python version is raised to 3.9+
         if sys.version_info < (3, 9):
             self.exit_on_error = True
 
@@ -1113,7 +1211,7 @@ class ArgumentParser(argparse.ArgumentParser):
             dict[str, dict[str, tuple[argparse.Action, str]]]: source to settings dict
         """
         # _source_to_settings is set in parse_know_args().
-        return self._source_to_settings  # type:ignore[attribute-error]
+        return self._source_to_settings  # type: ignore[attribute-error]
 
     def write_config_file(self, parsed_namespace, output_file_paths, exit_after=False):
         """Write the given settings to output files.
@@ -1258,9 +1356,7 @@ class ArgumentParser(argparse.ArgumentParser):
                     # --no-foo
                     args.append(action.option_strings[1])
             elif isinstance(action, argparse._CountAction):
-                for arg in args:
-                    if any([arg.startswith(s) for s in action.option_strings]):
-                        value = 0
+                # For count actions, repeat the flag the number of times specified
                 args += [action.option_strings[0]] * int(value)
             else:
                 self.error(
@@ -1428,7 +1524,7 @@ class ArgumentParser(argparse.ArgumentParser):
         for (
             source,
             settings,
-        ) in self._source_to_settings.items():  # type:ignore[argument-error]
+        ) in self._source_to_settings.items():  # type: ignore[argument-error]
             source = source.split("|")
             source = source_key_to_display_value_map[source[0]] % tuple(source[1:])
             r.write(source)
@@ -1599,8 +1695,9 @@ def already_on_command_line(
     )
 
 
-# TODO: Update to latest version of pydoctor when https://github.com/twisted/pydoctor/pull/414 has been merged
-# such that the alises can be documented automatically.
+# NOTE: Aliases below are not auto-documented. This could be improved by updating
+# to latest version of pydoctor when https://github.com/twisted/pydoctor/pull/414
+# has been merged, which would allow aliases to be documented automatically.
 
 # wrap ArgumentParser's add_argument(..) method with the one above
 argparse._ActionsContainer.original_add_argument_method = (

--- a/setup.py
+++ b/setup.py
@@ -2,7 +2,6 @@ import logging
 import os
 import sys
 
-
 try:
     from setuptools import setup
 except ImportError:

--- a/tests/test_configargparse.py
+++ b/tests/test_configargparse.py
@@ -1,4 +1,5 @@
 import argparse
+from collections import OrderedDict
 import configargparse
 from contextlib import contextmanager
 import inspect
@@ -17,6 +18,12 @@ if sys.version_info >= (3, 10):
     OPTIONAL_ARGS_STRING = "options"
 else:
     OPTIONAL_ARGS_STRING = "optional arguments"
+
+# In Python 3.13+, short option metavars are no longer repeated
+if sys.version_info >= (3, 13):
+    SHORT_OPT_METAVAR = ""  # e.g., "-g, --my-cfg-file MY_CFG_FILE"
+else:
+    SHORT_OPT_METAVAR = "(?: {metavar})?"  # e.g., "-g MY_CFG_FILE, --my-cfg-file MY_CFG_FILE" or "-g, --my-cfg-file MY_CFG_FILE"
 
 # set COLUMNS to get expected wrapping
 os.environ["COLUMNS"] = "80"
@@ -315,6 +322,9 @@ class TestBasicUseCases(TestCase):
         )
 
         if not use_groups:
+            short_g = SHORT_OPT_METAVAR.format(metavar="MY_CFG_FILE")
+            short_d = SHORT_OPT_METAVAR.format(metavar="DBSNP")
+            short_f = SHORT_OPT_METAVAR.format(metavar="FRMT")
             self.assertRegex(
                 self.format_help(),
                 "usage: .* \\[-h\\] --genome GENOME \\[-v\\] -g MY_CFG_FILE\n?"
@@ -325,13 +335,16 @@ class TestBasicUseCases(TestCase):
                 "  -h, --help \\s+ show this help message and exit\n"
                 "  --genome GENOME \\s+ Path to genome file\n"
                 "  -v\n"
-                "  -g(?: MY_CFG_FILE)?, --my-cfg-file MY_CFG_FILE\n"
-                "  -d(?: DBSNP)?, --dbsnp DBSNP\\s+\\[env var: DBSNP_PATH\\]\n"
-                "  -f(?: FRMT)?, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n\n"
+                f"  -g{short_g}, --my-cfg-file MY_CFG_FILE\n"
+                f"  -d{short_d}, --dbsnp DBSNP\\s+\\[env var: DBSNP_PATH\\]\n"
+                f"  -f{short_f}, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n\n"
                 % OPTIONAL_ARGS_STRING
                 + 7 * r"(.+\s*)",
             )
         else:
+            short_g = SHORT_OPT_METAVAR.format(metavar="MY_CFG_FILE")
+            short_d = SHORT_OPT_METAVAR.format(metavar="DBSNP")
+            short_f = SHORT_OPT_METAVAR.format(metavar="FRMT")
             self.assertRegex(
                 self.format_help(),
                 "usage: .* \\[-h\\] --genome GENOME \\[-v\\] -g MY_CFG_FILE\n?"
@@ -343,10 +356,10 @@ class TestBasicUseCases(TestCase):
                 "g1:\n"
                 "  --genome GENOME \\s+ Path to genome file\n"
                 "  -v\n"
-                "  -g(?: MY_CFG_FILE)?, --my-cfg-file MY_CFG_FILE\n\n"
+                f"  -g{short_g}, --my-cfg-file MY_CFG_FILE\n\n"
                 "g2:\n"
-                "  -d(?: DBSNP)?, --dbsnp DBSNP\\s+\\[env var: DBSNP_PATH\\]\n"
-                "  -f(?: FRMT)?, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n\n"
+                f"  -d{short_d}, --dbsnp DBSNP\\s+\\[env var: DBSNP_PATH\\]\n"
+                f"  -f{short_f}, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n\n"
                 % OPTIONAL_ARGS_STRING
                 + 7 * r"(.+\s*)",
             )
@@ -469,15 +482,18 @@ class TestBasicUseCases(TestCase):
             "  --format: \\s+ BED\n",
         )
 
+        short_f1 = SHORT_OPT_METAVAR.format(metavar="TYPE1_CFG_FILE")
+        short_f2 = SHORT_OPT_METAVAR.format(metavar="TYPE2_CFG_FILE")
+        short_f = SHORT_OPT_METAVAR.format(metavar="FRMT")
         self.assertRegex(
             self.format_help(),
             r"usage: .* \[-h\] --genome GENOME \[-v\]\s+\(-f1 TYPE1_CFG_FILE \|"
             r"\s+-f2 TYPE2_CFG_FILE\)\s+\(-f FRMT \| -b\)\n\n"
             "%s:\n"
             "  -h, --help            show this help message and exit\n"
-            "  -f1(?: TYPE1_CFG_FILE)?, --type1-cfg-file TYPE1_CFG_FILE\n"
-            "  -f2(?: TYPE2_CFG_FILE)?, --type2-cfg-file TYPE2_CFG_FILE\n"
-            "  -f(?: FRMT)?, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n"
+            f"  -f1{short_f1}, --type1-cfg-file TYPE1_CFG_FILE\n"
+            f"  -f2{short_f2}, --type2-cfg-file TYPE2_CFG_FILE\n"
+            f"  -f{short_f}, --format FRMT\\s+\\[env var: OUTPUT_FORMAT\\]\n"
             "  -b, --bam\\s+\\[env var: BAM_FORMAT\\]\n\n"
             "group1:\n"
             "  --genome GENOME       Path to genome file\n"
@@ -1054,12 +1070,13 @@ class TestMisc(TestCase):
         ns = self.parse("-c " + temp_cfg2.name)
         self.assertEqual(ns.genome, "hg20")
 
+        short_c = SHORT_OPT_METAVAR.format(metavar="CONFIG_FILE")
         self.assertRegex(
             self.format_help(),
             r"usage: .* \[-h\] -c CONFIG_FILE --genome GENOME\n\n"
             r"%s:\n"
             r"  -h, --help\s+ show this help message and exit\n"
-            r"  -c(?: CONFIG_FILE)?, --config CONFIG_FILE\s+ my config file\n"
+            rf"  -c{short_c}, --config CONFIG_FILE\s+ my config file\n"
             r"  --genome GENOME\s+ Path to genome file\n\n" % OPTIONAL_ARGS_STRING
             + 5 * r"(.+\s*)",
         )
@@ -1123,14 +1140,16 @@ class TestMisc(TestCase):
         self.add_arg("--arg1", help="Arg1 help text", required=True)
         self.add_arg("--flag", help="Flag help text", action="store_true")
 
+        short_c = SHORT_OPT_METAVAR.format(metavar="CONFIG_FILE")
+        short_w = SHORT_OPT_METAVAR.format(metavar="CONFIG_OUTPUT_PATH")
         self.assertRegex(
             self.format_help(),
             r"usage: .* \[-h\] -c CONFIG_FILE\s+"
             r"\[-w CONFIG_OUTPUT_PATH\]\s* --arg1\s+ARG1\s*\[--flag\]\s*"
             "%s:\\s*"
             "-h, --help \\s* show this help message and exit "
-            r"-c(?: CONFIG_FILE)?, --config CONFIG_FILE\s+my config file "
-            r"-w(?: CONFIG_OUTPUT_PATH)?, --write-config CONFIG_OUTPUT_PATH takes "
+            rf"-c{short_c}, --config CONFIG_FILE\s+my config file "
+            rf"-w{short_w}, --write-config CONFIG_OUTPUT_PATH takes "
             r"the current command line args and writes them "
             r"out to a config file at the given path, then exits "
             r"--arg1 ARG1 Arg1 help text "
@@ -1784,24 +1803,20 @@ class TestTomlConfigParser(unittest.TestCase):
         return f
 
     def test_section(self):
-        f = self.write_toml_file(
-            """
+        f = self.write_toml_file("""
             [section]
             key1 = 'toml1'
             key2 = 'toml2'
-            """
-        )
+            """)
         parser = configargparse.TomlConfigParser(["section"])
         self.assertEqual(parser.parse(f), {"key1": "toml1", "key2": "toml2"})
 
     def test_no_sections(self):
-        f = self.write_toml_file(
-            """
+        f = self.write_toml_file("""
             [section]
             key1 = 'toml1'
             key2 = 'toml2'
-            """
-        )
+            """)
         parser = configargparse.TomlConfigParser([])
         self.assertEqual(parser.parse(f), {})
 
@@ -1817,25 +1832,115 @@ class TestTomlConfigParser(unittest.TestCase):
 
     @unittest.expectedFailure  # Ints should be strings
     def test_advanced(self):
-        f = self.write_toml_file(
-            """
+        f = self.write_toml_file("""
             [tool.section]
             key1 = "toml1"
             key2 = [1, 2, 3]
-            """
-        )
+            """)
         parser = configargparse.TomlConfigParser(["tool.section"])
         self.assertEqual(parser.parse(f), {"key1": "toml1", "key2": ["1", "2", "3"]})
 
-    def test_fails_binary_read(self):
+    @unittest.skipIf(
+        sys.version_info < (3, 11),
+        "Binary mode only supported with tomllib (Python 3.11+)",
+    )
+    def test_binary_read_works(self):
+        # Binary mode now works with tomllib (Python 3.11+)
         f = self.write_toml_file(
             b"""[tool.section]\nkey1 = "toml1"
             """,
             obj=BytesIO,
         )
         parser = configargparse.TomlConfigParser(["tool.section"])
+        # Should successfully parse binary stream
+        self.assertEqual(parser.parse(f), {"key1": "toml1"})
+
+    @unittest.skipIf(
+        sys.version_info >= (3, 11),
+        "On Python 3.11+, tomllib handles binary; this tests the toml package fallback",
+    )
+    def test_binary_read_fails_without_tomllib(self):
+        # Without tomllib (Python < 3.11), binary streams should fail
+        f = self.write_toml_file(
+            b"""[section]\nkey1 = "toml1"\n""",
+            obj=BytesIO,
+        )
+        parser = configargparse.TomlConfigParser(["section"])
         with self.assertRaises(configargparse.ConfigFileParserException):
             parser.parse(f)
+
+    def test_serialize_with_section(self):
+        parser = configargparse.TomlConfigParser(["section"])
+        try:
+            import toml
+        except ImportError:
+            self.skipTest("toml package not installed")
+        items = OrderedDict([("key1", "val1"), ("key2", "val2")])
+        result = parser.serialize(items)
+        # Verify round-trip: parse the serialized output
+        parsed = parser.parse(StringIO(result))
+        self.assertEqual(parsed, {"key1": "val1", "key2": "val2"})
+
+    def test_serialize_nested_section(self):
+        parser = configargparse.TomlConfigParser(["tool.section"])
+        try:
+            import toml
+        except ImportError:
+            self.skipTest("toml package not installed")
+        items = OrderedDict([("key1", "val1")])
+        result = parser.serialize(items)
+        parsed = parser.parse(StringIO(result))
+        self.assertEqual(parsed, {"key1": "val1"})
+
+    def test_serialize_no_sections(self):
+        parser = configargparse.TomlConfigParser([])
+        try:
+            import toml
+        except ImportError:
+            self.skipTest("toml package not installed")
+        items = OrderedDict([("key1", "val1")])
+        result = parser.serialize(items)
+        self.assertIn("key1", result)
+
+    def test_serialize_without_toml_package(self):
+        parser = configargparse.TomlConfigParser(["section"])
+        items = OrderedDict([("key1", "val1")])
+        # Mock tomllib import succeeding but toml import failing
+        import unittest.mock as mock
+
+        with mock.patch.dict(sys.modules, {"toml": None}):
+            with self.assertRaises(configargparse.ConfigFileParserException):
+                parser.serialize(items)
+
+
+class TestIniConfigParserSerialize(unittest.TestCase):
+    def test_serialize_basic(self):
+        parser = configargparse.IniConfigParser(["section"], False)
+        items = OrderedDict([("key1", "val1"), ("key2", "val2")])
+        result = parser.serialize(items)
+        # Verify round-trip
+        parsed = parser.parse(StringIO(result))
+        self.assertEqual(parsed, {"key1": "val1", "key2": "val2"})
+
+    def test_serialize_list_as_python_syntax(self):
+        parser = configargparse.IniConfigParser(["section"], False)
+        items = OrderedDict([("key1", ["a", "b", "c"])])
+        result = parser.serialize(items)
+        parsed = parser.parse(StringIO(result))
+        self.assertEqual(parsed, {"key1": ["a", "b", "c"]})
+
+    def test_serialize_list_multiline(self):
+        parser = configargparse.IniConfigParser(["section"], True)
+        items = OrderedDict([("key1", ["a", "b", "c"])])
+        result = parser.serialize(items)
+        # Should use multiline format
+        self.assertIn("\n", result.split("key1")[1])
+
+    def test_serialize_default_section(self):
+        parser = configargparse.IniConfigParser([], False)
+        items = OrderedDict([("key1", "val1")])
+        result = parser.serialize(items)
+        self.assertIn("key1", result)
 
 
 class TestCompositeConfigParser(unittest.TestCase):
@@ -1942,6 +2047,24 @@ class TestCompositeConfigParser(unittest.TestCase):
         self.write_toml_file_extra()
         with self.assertRaises(SystemExit):
             self.parser.parse_args([])
+
+    def test_composite_serialize_delegates_to_first_parser(self):
+        composite = configargparse.CompositeConfigParser(
+            [
+                configargparse.IniConfigParser(["section"], False),
+                configargparse.DefaultConfigFileParser,
+            ]
+        )()
+        items = OrderedDict([("key1", "val1"), ("key2", "val2")])
+        result = composite.serialize(items)
+        # Should use INI format (first parser)
+        self.assertIn("[section]", result)
+        self.assertIn("key1 = val1", result)
+
+    def test_composite_serialize_empty_parsers(self):
+        composite = configargparse.CompositeConfigParser([])()
+        with self.assertRaises(configargparse.ConfigFileParserException):
+            composite.serialize(OrderedDict())
 
 
 ################################################################################

